src/mlpack/methods/ann/layer/neural_turing_machine.hpp

```diff
@@ -0,0 +1,256 @@
+   */
+  void ResetCell();
+
+  //! The value of the deterministic parameter.
+  bool Deterministic() const { return deterministic; }
+  //! Modify the value of the deterministic parameter.
+  bool& Deterministic() { return deterministic; }
+
+  //! Get the parameters.
+  OutputDataType const& Parameters() const { return weights; }
+  //! Modify the parameters.
+  OutputDataType& Parameters() { return weights; }
+
+  //! Get the input parameter.
+  InputDataType const& InputParameter() const { return inputParameter; }
+  //! Modify the input parameter.
+  InputDataType& InputParameter() { return inputParameter; }
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return gradient; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return gradient; }
+
+  //! Get the model modules.
+  std::vector<LayerTypes>& Model() { return network; }
+
+  /**
+   * Serialize the layer
+   */
+  template<typename Archive>
+  void Serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored number of input units.
+  size_t inSize;
+
+  //! Locally-stored number of output units.
+  size_t outSize;
+
+  //! Number of memory locations.
+  size_t numMem;
+
+  //! Size of memory locations.
+  size_t memSize;
+
+  //! Circular shift rotation size.
+  size_t shiftSize;
+
+  //! Memory read layer.
+  LayerTypes readMem;
+
+  //! Memory write layer.
+  LayerTypes writeMem;
+
+  //! Controller.
+  LayerTypes controller;
+
+  //! Locally stored memory content error.
+  arma::mat dMem;
+
+  //! Locally stored previous memory content error.
+  arma::mat dMemPrev;
+
+  //! All zeros vector for initializing read.
+  arma::mat allZeros;
+
+  //! All ones matrix for initializing memory.
+  arma::mat allOnes;
+
+  //! Storing all the memory contents for backward pass.
+  std::list<arma::mat> memoryHistory;
+
+  //! Backward pass iterator to stored memory content.
+  std::list<arma::mat>::iterator bMemoryHistory;
+
+  //! Storing all the memory reads for backward pass.
+  std::list<arma::mat> lReads;
+
+  //! Backward pass iterator to stored memory reads.
+  std::list<arma::mat>::iterator gReads;
+
+  //! Locally stored memory read error.
+  arma::mat dRead;
+
+  //! Locally stored memory write error.
+  arma::mat dWriteHead;
+
+  //! Locally-stored weight object.
+  OutputDataType weights;
+
+  //! Locally-stored output parameter visitor.
+  OutputParameterVisitor outputParameterVisitor;
+
+  //! Locally-stored delta visitor.
+  DeltaVisitor deltaVisitor;
+
+  //! Locally-stored list of network modules.
+  std::vector<LayerTypes> network;
```

Comment:
Ah, sorry for the slow response.  You are right, since it is a boost::variant to a pointer, I think it will not be a problem to also store the other objects.